### ISSUE TITLE
Share/collect  non-tensor objects between processes.

### ISSUE THREAD

**Issue by dalek-who** [Lightning-AI/lightning#14362]
### Motivation

Sometimes we need to share/collect some non-tensor objects between process in DDP mode, mainly includes two kinds of scenes:
- **share status**: for example, a random port to connect database, a timestamp when the experiment starts, an auto-created output directory, etc. These status are created during running and cannot be assigned manually.
- **collect outputs**: for example, in test/predict, collect all outputs to `gloabl_rank_zero` process  in `on_{test,preidct}_{batch,epoch}_end` to write predict results and compute metrics over the full dataset.

Current functions cannot satisfy these requests:
- `LightningModule.all_gather` can only deal with tensors. Other objects (like string, id list, Path ...) cannot be gathered or shared.
-  `DP` or `DDPSpawn`: they are not as "convenient" as DDP. DP is not encouraged by torch, and DDPSpawn sometimes cannot be started.
- `torch.multiprocessing.SimpleQueue`: there are no tutorials on how to use it with pytorch-lightning DDP mode.

### Alternatives

Consider something like this:
```python
def collect_all_objects(obj: object, global_zero_only: bool=False, default=None):
    ...
    objects_from_process = {
        0: obj_0,
        1: obj_1,
        global_rank_n: obj_n,
    }
    if not trainer.is_global_zero and global_zero_only:
        return default
    else:
        return objects_from_process
```
 This function can also be used to share status between processes: each process collect status and only use the status from rank_0.

**Important:**
When using DDP, the sampler may padding some examples in the last batch and cause the following problem:
- #11102

In this case, manually deduplication is necessary after collecting objects. However, deduplication is much more easier than dealing with inter-process synchronization.

### Additional context

There are some questions related to these requests:

1.**share status**:
- [Share state between DDP processes](https://forums.pytorchlightning.ai/t/share-state-between-ddp-processes/963)
- [Sync output dir between DDP processes](https://forums.pytorchlightning.ai/t/sync-output-dir-between-ddp-processes/779)

2.**collect outputs**:
-  #5788
-  #10618
- #11086
- #9259

**Comment by dalek-who** [comment]
Currently, I have a heavy solution for DDP multi-gpu test, using an temporary database ([ZODB](https://zodb.org/en/latest/) and [ZEO](https://zeo.readthedocs.io/en/latest/)) as a shared buffer:
```python
class MyModule(pl. LightningModule):

def prepare_data(self) -> None:
    # global_rank_0 start an in-memory ZEO server. ip/port are provided from command-line arguments.
    # create a btree as buffer

def test_epoch_end(self, test_step_outputs) -> None:
    # each process starts a ZEO client
    # each process write `test_step_outputs` to ZODB database (btree)

    self.trainer.strategy.barrier()
    
    if not self.trainer.is_global_zero:
        return
    else:  # global_zero
        # read all results from ZODB database
        # deduplication (each of my example has a unique id)
        # write results
        # compute performance

```

This solution is useful but too heavy.

**Comment by dalek-who** [comment]
@awaelchli Thanks, finally I get an elegant solution with [torch.distributed.all_gather_object](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather_object). Here is my code:
```python 
from typing import List, Dict, Tuple, NamedTuple, Union, Optional, Any, Type, TypeVar
import torch.distributed as dist
from pytorch_lightning.utilities.distributed import group as _group

T = TypeVar("T")

class MyModule(pl. LightningModule):

    def test_epoch_end(self, test_step_outputs: Union[List[Dict], List[List[Dict]]]) -> None:
        list_gather_step_outputs = self._gather_objects(obj=test_step_outputs)
        if not self.trainer.is_global_zero:
            return
        else:  # global zero only
            ...  # important!!! excluded duplicated outputs!
            ...  # other things, like save result, compute performance, etc.
        
    def _gather_objects(self, obj: T) -> Optional[List[T]]:
        if not self.trainer.is_global_zero:
            dist.gather_object(obj=obj, object_gather_list=None, dst=0, group=_group.WORLD)
            return None
        else: # global-zero only
            list_gather_obj = [None] * self.trainer.world_size   # the container of gathered objects.
            dist.gather_object(obj=obj, object_gather_list=list_gather_obj, dst=0, group=_group.WORLD)
            return list_gather_obj
        
```

I think adding the solution in the document and refering readers to [torch.distributed](https://pytorch.org/docs/stable/distributed.htm)  is better than wrapping it in `self.all_gather`. Here's my reasons:
1. This solution is simple enough. 
2.`torch.distributed` provides more DDP communication functions than `gather_object`, like `broadcast_object_list`, `all_gather_object`, `scatter_object_list`, `Distributed Key-Value Store`, etc.

**Comment by Borda** [comment]
@dalek-who having this case in docs would be great!

**Comment by dalek-who** [comment]
**Notice:** 
@Borda The code I provided is for `pytorch-lightning==1.6` and is now outdated. For `pytorch-lightning==1.9.5`, 
> from pytorch_lightning.utilities.distributed import group as _group

will causes an ImportError. I need `_group.WORLD` in `dist.gather_object(..., group=_group.WORLD)`. Could it be replaced with the following code?
> from torch.distributed import group as _group
